python/fnllm/fnllm/openai/factories/chat.py

```diff
@@ -0,0 +1,103 @@
+# Copyright (c) 2024 Microsoft Corporation.
+
+"""Factory functions for creating OpenAI LLMs."""
+
+from fnllm.caching.base import Cache
+from fnllm.events.base import LLMEvents
+from fnllm.limiting.base import Limiter
+from fnllm.openai.config import OpenAIConfig
+from fnllm.openai.llm.chat import OpenAIChatLLMImpl
+from fnllm.openai.llm.chat_streaming import OpenAIStreamingChatLLM
+from fnllm.openai.llm.chat_text import OpenAITextChatLLM
+from fnllm.openai.llm.features.tools_parsing import OpenAIParseToolsLLM
+from fnllm.openai.llm.services.history_extractor import OpenAIHistoryExtractor
+from fnllm.openai.llm.services.json import create_json_handler
+from fnllm.openai.llm.services.usage_extractor import OpenAIUsageExtractor
+from fnllm.openai.types.client import (
+    OpenAIChatLLM,
+    OpenAIClient,
+    OpenAIStreamingChatLLMInstance,
+    OpenAITextChatLLMInstance,
+)
+from fnllm.services.cache_interactor import CacheInteractor
+from fnllm.services.variable_injector import VariableInjector
+
+from .client import create_openai_client
+from .utils import create_limiter, rate_limiter, retryer
+
+
+def create_openai_chat_llm(
+    config: OpenAIConfig,
+    *,
+    client: OpenAIClient | None = None,
+    cache: Cache | None = None,
+    events: LLMEvents | None = None,
+) -> OpenAIChatLLM:
+    """Create an OpenAI chat LLM."""
+    if client is None:
+        client = create_openai_client(config)
+
+    limiter = create_limiter(config)
+
+    text_chat_llm = _create_openai_text_chat_llm(
+        client=client,
+        config=config,
+        cache=cache,
+        events=events,
+        limiter=limiter,
+    )
+    streaming_chat_llm = _create_openai_streaming_chat_llm(
+        client=client,
+        config=config,
+        events=events,
+        limiter=limiter,
+    )
+    return OpenAIChatLLMImpl(
+        text_chat_llm=text_chat_llm,
+        streaming_chat_llm=streaming_chat_llm,
```

Comment:
Do we need both? Is it a common scenario where we have a single instance with both implementations being used?

Reply:
Yeah, so if you look in `python/fnllm/fnllm/openai/llm/chat.py`, you'll see that there's some overloads. It will delegate to the streaming endpoint if you call the llm with `stream=True`, otherwise, by default it uses the chat llm